### PR TITLE
Add a validating webhook to EdgeDevice

### PR DESCRIPTION
**Pull request by rgolangh**
Prevent irrelevant or conflicting device specifications.

Signed-off-by: Roy Golan <rgolan@redhat.com>

ECOPROJECT-347

### CODE REVIEW

File: api/v1alpha1/edgedevice_webhook.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	edgedevicelog.Info("validate update", "name", d.Name)
+	err := d.validateCreateOrUpdate()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *EdgeDevice) validateCreateOrUpdate() error {
+	if d.Spec.Storage.S3.CreateOBC &&
```
Comment (masayag) [inline api/v1alpha1/edgedevice_webhook.go:70]: iinw, any of `.Storage.S3.CreateOBC` can be nil, therefore need to be cautions not to address it without checking each level

File: api/v1alpha1/edgedevice_webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+				device.Spec.Storage.S3.ConfigMapName = "cm"
+				device.Spec.Storage.S3.CreateOBC = true
+				Expect(device.ValidateCreate()).To(HaveOccurred())
+			})
+			It("should fail to update", func() {
+				// given
+				device.Spec.Storage.S3.SecretName = "secret"
+				device.Spec.Storage.S3.ConfigMapName = "cm"
+				device.Spec.Storage.S3.CreateOBC = true
+				Expect(device.ValidateUpdate(nil)).To(HaveOccurred())
```
Comment (jakub-dzon) [inline api/v1alpha1/edgedevice_webhook_test.go:43]: Please add new line and `// then` before `Expect`

File: config/webhook/manifests.yaml [hunk 3]
```diff
@@ -1,4 +1,33 @@
 
+---
+apiVersion: admissionregistration.k8s.io/v1
+kind: MutatingWebhookConfiguration
```
Comment (jakub-dzon) [inline config/webhook/manifests.yaml:4]: There is no mutation in the webhook - please remove.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: PROJECT [hunk 4]
```diff
@@ -16,6 +16,9 @@ resources:
   kind: EdgeDevice
   path: github.com/project-flotta/flotta-operator/api/v1alpha1
   version: v1alpha1
+  webhooks:
```
Comment (masayag) [inline PROJECT:19]: although not part of this PR, can you add the same for EdgeDeployment below?

File: config/webhook/manifests.yaml [hunk 5]
```diff
@@ -26,3 +26,24 @@ webhooks:
     resources:
     - edgedeployments
   sideEffects: None
+- admissionReviewVersions:
```
Comment (masayag) [inline config/webhook/manifests.yaml:29]: can we reuse the same service for both kinds and unified it so the only changes to this file is:
1. Generalize the name of the path of the service in line 15 (e.g. s//validate-management-project-flotta-io-v1alpha1-edgedeployment//validate-management-project-flotta-io-v1alpha1)
2. Extend the rules to refer to edgedevices as well

File: PROJECT (hunk 4 above)
Comment (rgolangh) [reply PROJECT:19]: done

File: config/webhook/manifests.yaml (hunk 5 above)
Comment (rgolangh) [reply config/webhook/manifests.yaml:29]: any particular reason for reusing it? the different section will be needed if we would want various config for each.

File: config/default/manager_webhook_patch.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+apiVersion: apps/v1
```
[the remark this answers is not in the document]
Comment (rgolangh) [reply config/default/manager_webhook_patch.yaml:1]: this is all operator sdk that generated that. I wonder how the initial webhook was created . 
removing this whole section

File: main.go [hunk 7]
```diff
@@ -231,6 +233,10 @@ func main() {
 		}
 	}
 
+	if err = (&managementv1alpha1.EdgeDevice{}).SetupWebhookWithManager(mgr); err != nil {
```
Comment (rgolangh) [inline main.go:236]: I'll move this under the condition in 229. it looks intentional :)

File: config/webhook/manifests.yaml (hunk 5 above)
Comment (masayag) [reply config/webhook/manifests.yaml:29]: no, let's keep it separated.

File: config/webhook/manifests.yaml [hunk 8]
```diff
@@ -26,3 +26,24 @@ webhooks:
...
+  - v1
+  - v1beta1
+  clientConfig:
+    service:
+      name: webhook-service
+      namespace: system
+      path: /validate-management-project-flotta-io-v1alpha1-edgedevice
+  failurePolicy: Fail
+  name: vedgedevice.kb.io
```
Comment (masayag) [inline config/webhook/manifests.yaml:38]: pls update the name to the CR name edgedevice.management.project-flotta.io

**Comment by jakub-dzon** [comment]
@rgolangh - how is the work going on that feature? It seems to be needed for #148.

**Comment by pkliczewski** [comment]
@rgolangh will you be able to finalize this PR?

**Review by jakub-dzon — APPROVED** [review]

**Review by masayag — APPROVED** [review]
